ACP-196: Precompile for secp256r1 Curve Support (#204)
avalanche-foundation/ACPs: ACPs/204-precompile-secp256r1/README.md

This ACP proposes implementing EIP-7212's secp256r1 precompiled contract to unlock significant ecosystem benefits:

### **Enterprise & Institutional Adoption**
- **Reduced onboarding friction**: Enterprises can leverage existing biometric authentication infrastructure instead of managing seed phrases or hardware wallets
- **Regulatory compliance**: Institutions can utilize their approved device security standards and identity management systems
- **Cost optimization**: 100x gas reduction (from 200k-330k to 3,450 gas) makes enterprise-scale applications economically viable

The 100x gas cost reduction makes these use cases economically viable while maintaining the security properties institutions and users expect from their existing devices.

## Specification

> This ACP implements [EIP-7212](https://eips.ethereum.org/EIPS/eip-7212) for secp256r1 signature verification on Avalanche. The specification follows EIP-7212 exactly, with the precompiled contract deployed at address `0x0200000000000000000000000000000000000006`.

Reviewer: ```suggestion
This ACP implements [EIP-7212](https://eips.ethereum.org/EIPS/eip-7212) for secp256r1 signature verification on Avalanche. The specification follows EIP-7212 exactly, with the precompiled contract deployed at address `0x0200000000000000000000000000000000000006`, imported to coreth via the precompile [registry](https://github.com/ava-labs/coreth/blob/master/precompile/registry/registry.go).
```
The registry's comments will include the [deployed address listed](https://github.com/ava-labs/coreth/pull/990/files) if the precompile is added.

2. No existing opcodes or consensus rules are modified
3. The change is additive and opt-in for applications

Adoption requires a coordinated network upgrade for the C-Chain. Subnet-EVM chains can adopt this enhancement independently by upgrading their client software.

## Security Considerations

### Cryptographic Security
- The secp256r1 curve is standardized by NIST and widely vetted
- Security properties are comparable to secp256k1 (used by ECRECOVER)
- Implementation follows NIST FIPS 186-3 specification exactly

### Implementation Security  
- Signature verification (not recovery) approach maximizes compatibility with existing P-256 ecosystem
- No malleability check included to match NIST specification, but wrapper libraries should add this
- Input validation prevents invalid curve points and out-of-range signature components

### Network Security
- Gas cost prevents potential DoS attacks through expensive computation
- Precompile implementation should use battle-tested cryptographic libraries
- No consensus-level security implications beyond standard precompile considerations

## Reference Implementation

The implementation will build upon existing work:

> 1. **EIP-7212 Reference**: The [BOR implementation](https://github.com/maticnetwork/bor/pull/1069) of EIP-7212 provides the foundation
> 2. **Avalanche Integration**: 
>    - **coreth**: Integration with Avalanche's C-Chain (Avalanche's fork of go-ethereum)
> 3. **Cryptographic Library**: Implementation utilizes Go's standard library `crypto/ecdsa` and `crypto/elliptic` packages, which implement NIST P-256 per FIPS 186-3 ([Go documentation](https://pkg.go.dev/crypto/elliptic#P256))

Reviewer: ```suggestion
1. EIP-7212 Reference: The [BOR implementation](https://github.com/maticnetwork/bor/pull/1069) of EIP-7212 provides the foundation
2. Coreth Implementation: Integration with Avalanche's C-Chain (Avalanche's fork of go-ethereum)
3. Cryptographic Library: Implementation utilizes Go's standard library `crypto/ecdsa` and `crypto/elliptic` packages, which implement NIST P-256 per FIPS 186-3 ([Go documentation](https://pkg.go.dev/crypto/elliptic#P256))
```

For complete technical specifications, validation requirements, and implementation details, refer to [RIP-7212](https://github.com/ethereum/RIPs/blob/master/RIPS/rip-7212.md).

## Backwards Compatibility

This ACP introduces a new precompiled contract and does not modify existing functionality. No backwards compatibility issues are expected since:

1. The precompile uses a previously unused address
2. No existing opcodes or consensus rules are modified
3. The change is additive and opt-in for applications

Adoption requires a coordinated network upgrade for the C-Chain. Other EVM L1s can adopt this enhancement independently by upgrading their client software.

## Security Considerations

### Cryptographic Security

- The secp256r1 curve is standardized by NIST and widely vetted
- Security properties are comparable to secp256k1 (used by ECRECOVER)
- Implementation follows NIST FIPS 186-3 specification exactly

### Implementation Security
 
- Signature verification (vs public-key recovery) approach maximizes compatibility with existing P-256 ecosystem

> - No malleability check included to match NIST specification, but wrapper libraries should add this

Reviewer 2: I feel like documenting an expectation of wrapper libraries to check this is a bit odd.
```suggestion
- No malleability check included to match NIST specification, but wrapper libraries may choose to add this
```
fwiw I don't think there is a problem with allowing malleability here. Just seems like a stronger statement than we should make.
